### ISSUE TITLE
Extend `PayloadAttributes` with configurable `gas_limit`

### ISSUE THREAD

**Issue by ralexstokes** [ethereum/builder-specs#29]
If we refer to the current execution Engine API spec for the build inputs: https://github.com/ethereum/execution-apis/blob/main/src/engine/specification.md#payloadattributesv1

we see that there is no way to specify a gas limit.

It is critical for the security of the network that this parameter remains under the control of the validator set and *not* the builder set (which should be much smaller and possibly not as aligned with ethereum overall).

I think an easy fix is to specify a `PayloadAttributesV2` message (h/t @mkalinin ) that simply appends a `gas_limit` field:

```
* gasLimit: QUANTITY, 64 Bits - value for the block gas limit of the new payload
```

**Comment by mkalinin** [comment]
I think it is reasonable for a validator to be capable of controlling the `gas_limit`. But only in the case when VC operator clearly understands what effect does this parameter may take and how can it affect the network. Considering this, I would make `gas_limit` configuration and Engine API parameter optional. With having it optional not experienced stakers could rely on EL and avoid misconfigurations that are potentially dangerous to the network.

Based on what @ralexstokes has proposed above, Engine API changes may look as follows:
```md
### PayloadAttributesV2
- `timestamp`: `QUANTITY`, 64 Bits - value for the `timestamp` field of the new payload
- `prevRandao`: `DATA`, 32 Bytes - value for the `prevRandao` field of the new payload
- `suggestedFeeRecipient`: `DATA`, 20 Bytes - suggested value for the `feeRecipient` field of the new payload
- `gasLimit`: `QUANTITY`, 64 Bits - value for the block gas limit of the new payload

### engine_forkchoiceUpdatedV2

#### Request

* method: "engine_forkchoiceUpdatedV2"
* params: 
  1. `forkchoiceState`: `Object` - instance of [`ForkchoiceStateV1`](#ForkchoiceStateV1)
  2. `payloadAttributes`: `Object|null` - instance of [`PayloadAttributesV2`](#PayloadAttributesV2) or `null`
* timeout: 8s
```

This change could be deployed as a soft fork in two steps: i) all EL clients release a support of this feature, nodes are upgraded ii) CL clients are released.

I feel like we really need to think about versioning in Engine API and make CL follow Postel's law (echoing @micahzoltu). This would probably require additional `status` message but would make soft fork deployment smooth.

**Comment by MicahZoltu** [comment]
The execution layer is (presumably) under control of validators, and the execution layer having control doesn't automatically mean that builders control the value.  I think we would be better off here just communicating with execution client devs that they shouldn't be letting builders decide gas limit.

**Comment by mkalinin** [comment]
@ralexstokes How is this change related to `gas_limit` parameter in the validator registration message of Builders API?

**Comment by ralexstokes** [comment]
@mkalinin @MicahZoltu sorry for not providing the context here:

if we add this parameter then we will be able to re-use the Engine API in the context of the external builder network.

currently: 

builders in this network acquire the gas limits via the `ValidatorRegistration`s and they *MUST* follow the validator's preference here when they are building a block for them to proposer. it is not specified how the builder will go about building this block and it is assumed they will be using custom software to do so.

with this proposal:
if we extend the Engine API so that it includes *all* of the building parameters then external block builders can re-use off-the-shelf building software (e.g. geth) without having to develop custom software (or at least as much custom software). this lowers the barrier to entry, keeping the builder market in healthy competition.

note:
- builders can still get validator pref's ahead of time and use custom software to know about the gas limit they should be targeting before they would be able know based on the implied timing of the Engine API (which is "just in time")
- this parameter does not replace the `gas_limit` in the `ValidatorRegistration`, it simply supplements it
- this parameter does not directly influence any **local** building a proposer may be doing *concurrent* with an external block pipeline (note the status quo is likely just running `geth` which has a single, process-wide gas limit parameter -- if you have multiple validators using one EL then there is also no easy way for them to have differing gas limit preferences but this is not a big deal as we can assume in this setting all of the validators are controlled by one logical entity who wishes to have a uniform gas limit)

**Comment by zah** [comment]
My impression is that there is a wide-spread confusion in the community regarding how this parameter should be configured.

It would be useful if someone tries to enumerate the various rational (and perhaps adversarial) strategies for setting the value of the parameter. The honest validator document present in this repo can potentially provide some guidelines as well.

**Comment by MicahZoltu** [comment]
TL;DR: set it to 30,000,000 and if there is an active DoS attack against the network that can be mitigated by lowering the gas limit (this will likely be communicated publicly by Ethereum Core developers), then update your configuration at that time to lower it until the developers can get patches out.

**Comment by nflaig** [comment]
the (target) gas limit is fully controlled by validators today and will be made even stricter with epbs/gloas as bids that don't adhere to the target gas limit as configured by the proposer will be ignored

the configuration is now fully on the validator client side with `PayloadAttributesV4` meaning operators no longer need to configure it on both the CL and EL side, the config on the validator client applies to both local payloads and bids from builders
- https://github.com/ethereum/execution-apis/pull/796

going to close this, feel free to reopen if there is still something unresolved here
